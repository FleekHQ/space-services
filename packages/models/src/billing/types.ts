import { AppTableItem } from '../types';

export interface Wallet {
  // The uuid of the owner
  ownerUuid: string;

  // The derived hash of the key or address of this wallet
  keyHash: string;

  // The amount of credits available in this wallet
  credits: number;

  // Date of creation in ISO format
  createdAt: string;
}

export interface RawWallet extends AppTableItem {
  // The amount of credits available in this wallet
  credits: number;

  // Date of creation in ISO format
  createdAt: string;
}

export interface StripeInfo {
  // Stripe customer e-mail address
  email: string;

  // Stripe customer id
  stripeCustomerId: string;

  // Date of creation in ISO format
  createdAt: string;
}

export interface RawStripeInfo extends AppTableItem {
  // Stripe customer id
  stripeCustomerId: string;

  // Date of creation in ISO format
  createdAt: string;
}

export interface StripeSubscription {
  // Stripe subscription id
  id: string;

  // Stripe customer id
  stripeCustomerId: string;

  // Date of creation in ISO format
  createdAt: string;

  // product key
  uuid: string;
}

export interface RawStripeSubscription extends AppTableItem {
  // Stripe customer id
  stripeCustomerId: string;

  // Date of creation in ISO format
  createdAt: string;
}
