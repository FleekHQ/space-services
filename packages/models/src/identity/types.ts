import { AppTableItem } from '../types';

export interface CreateIdentityInput {
  // The public key of the identity
  publicKey: string;

  // Human readable identifier
  username: string;
}

export interface Identity {
  // The public key of the identity
  publicKey: string;

  // Hexadecimal string that is derived from a public key
  address: string;

  // Human readable identifier
  username: string;

  // Date of creation in ISO format
  createdAt: string;

  // TODO: Twitter handle or other identity providers
}

export interface RawIdentity extends AppTableItem {
  // The public key of the identity
  publicKey: string;

  // Hexadecimal string that is derived from a public key
  address: string;

  // Human readable identifier
  username: string;

  // Date of creation in ISO format
  createdAt: string;
}

export interface CreateProofInput {
  // The type of proof
  proofType: ProofType;

  // The value of the proof. E.g. if this is an email proof, the value would be the actual email
  value: string;

  // The username of the user this proof belongs to
  username: string;
}

export enum ProofType {
  email = 'email',
}

export interface Proof {
  // The type of proof
  proofType: ProofType;

  // The value of the proof. E.g. if this is an email proof, the value would be the actual email
  value: string;

  // The username of the user this proof belongs to
  username: string;

  // Date of creation in ISO format
  createdAt: string;
}

export interface RawProof extends AppTableItem {
  // The type of proof
  proofType: ProofType;

  // The value of the proof. E.g. if this is an email proof, the value would be the actual email
  value: string;

  // The username of the user this proof belongs to
  username: string;

  // Date of creation in ISO format
  createdAt: string;
}
