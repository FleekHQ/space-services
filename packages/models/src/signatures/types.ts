import { AppTableItem } from '../types';

export interface Signature {
  /**
   * Base32 encoded public key
   */
  publicKey: string;

  /**
   * Challenge signed by private key
   */
  signature: string;
}

export interface RawSignature extends AppTableItem {
  publicKey: string;
  signature: string;
}
