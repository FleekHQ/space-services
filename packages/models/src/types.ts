export interface AppTableItem {
  pk: string;
  sk: string;
  gs1pk?: string;
  gs1sk?: string;
}
