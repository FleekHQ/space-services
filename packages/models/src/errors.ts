/* eslint-disable max-classes-per-file */

export class DomainError extends Error {
  constructor(message) {
    super(message);
    // Ensure the name of this error is the same as the class name
    this.name = this.constructor.name;
    // This clips the constructor invocation from the stack trace.
    // It's not absolutely essential, but it does make the stack trace a little nicer.
    //  @see Node.js reference (bottom)
    Error.captureStackTrace(this, this.constructor);
  }
}

export class ValidationError extends DomainError {
  constructor(message) {
    super(`Validation Error: ${message}`);
  }
}

export class NotFoundError extends DomainError {
  constructor(message) {
    super(`Not Found Error: ${message}`);
  }
}
