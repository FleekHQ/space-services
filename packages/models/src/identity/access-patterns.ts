import {
  IdentityRecord,
  RawIdentityRecord,
  ProofRecord,
  RawProofRecord,
  UsernameRecord,
  RawUsernameRecord,
  RawAddressRecord,
  AddressRecord,
} from './types';
import { PrimaryKey } from '../types';

// Note: Read ./README.md for details on access patterns

export const IDENTITY_KEY = 'id';
export const PROOF_KEY = 'proof';

export const getIdentityPrimaryKey = (uuid: string): PrimaryKey => ({
  pk: uuid,
  sk: IDENTITY_KEY,
});

export const getProofPrimaryKey = (
  value: string,
  type: string
): PrimaryKey => ({
  pk: `${PROOF_KEY}#${value}`,
  sk: type,
});

export const getUsernamePrimaryKey = (username: string): PrimaryKey => ({
  pk: `u#${username}`,
  sk: `username`,
});

export const getAddressPrimaryKey = (address: string): PrimaryKey => ({
  pk: address,
  sk: `address`,
});

export const mapIdentityDbObject = (
  identity: IdentityRecord
): RawIdentityRecord => {
  const { uuid, ...rest } = identity;

  return {
    ...getIdentityPrimaryKey(uuid),
    ...rest,
  };
};

export const parseDbObjectToIdentity = (
  dbObject: RawIdentityRecord
): IdentityRecord => ({
  uuid: dbObject.pk,
  address: dbObject.address,
  username: dbObject.username,
  displayName: dbObject.displayName,
  publicKey: dbObject.publicKey,
  createdAt: dbObject.createdAt,
  avatarUrl: dbObject.avatarUrl,
});

export const mapProofDbObject = (proof: ProofRecord): RawProofRecord => {
  return {
    ...getProofPrimaryKey(proof.value, proof.type),
    gs1pk: proof.uuid,
    gs1sk: `${PROOF_KEY}#${proof.type}`,
    ...proof,
  };
};

export const parseDbObjectToProof = (
  dbObject: RawProofRecord
): ProofRecord => ({
  value: dbObject.value,
  uuid: dbObject.uuid,
  type: dbObject.type,
  createdAt: dbObject.createdAt,
});

export const mapUsernameDbObject = (
  input: UsernameRecord
): RawUsernameRecord => {
  return {
    ...getUsernamePrimaryKey(input.username),
    gs1pk: input.uuid,
    gs1sk: `username`,
    createdAt: input.createdAt,
  };
};

export const parseDbObjectToUsername = (
  dbObject: RawUsernameRecord
): UsernameRecord => ({
  uuid: dbObject.gs1pk,
  username: dbObject.pk.split(`#`).pop(),
  createdAt: dbObject.createdAt,
});

export const mapAddressDbObject = (input: AddressRecord): RawAddressRecord => {
  return {
    ...getAddressPrimaryKey(input.address),
    gs1pk: input.uuid,
    gs1sk: `address`,
    createdAt: input.createdAt,
  };
};

export const parseDbObjectToAddress = (
  dbObject: RawAddressRecord
): AddressRecord => ({
  uuid: dbObject.gs1pk,
  address: dbObject.pk,
  createdAt: dbObject.createdAt,
});
