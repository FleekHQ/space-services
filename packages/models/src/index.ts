export { SignatureModel } from './signatures';
export { IdentityModel } from './identity';
export { VaultModel } from './vault';
export { default as createNotificationsModel } from './notifications';
export { AccountWithBilling, default as createBillingModel } from './billing';
export * from './errors';
export default {};
