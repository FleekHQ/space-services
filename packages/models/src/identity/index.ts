import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import { deriveAddressFromPubKey } from '@packages/crypto';
import { v4 as uuidv4 } from 'uuid';
import {
  mapIdentityDbObject,
  mapProofDbObject,
  parseDbObjectToIdentity,
  mapUsernameDbObject,
  mapAddressDbObject,
  getAddressPrimaryKey,
  parseDbObjectToAddress,
  getIdentityPrimaryKey,
  getUsernamePrimaryKey,
  parseDbObjectToUsername,
} from './access-patterns';
import {
  CreateIdentityInput,
  IdentityRecord,
  CreateProofInput,
  ProofRecord,
  RawIdentityRecord,
  RawAddressRecord,
} from './types';
import { validateIdentity } from './validations';
import { BaseModel } from '../base';
import { NotFoundError, ValidationError } from '../errors';

export class IdentityModel extends BaseModel {
  constructor(env: string, client: DocumentClient = new DocumentClient()) {
    const table = `space_table_${env}`;
    super(table, client);
  }

  public async createIdentity(
    input: CreateIdentityInput
  ): Promise<IdentityRecord> {
    const createdAt = new Date(Date.now()).toISOString();
    const address = deriveAddressFromPubKey(input.publicKey);

    const newIdentity = {
      uuid: uuidv4(),
      address,
      createdAt,
      publicKey: input.publicKey,
      username: input.username,
    };

    await validateIdentity(this, newIdentity);

    const dbItem = mapIdentityDbObject(newIdentity);

    // reserve username for this identity
    if (newIdentity.username) {
      await this.put(mapUsernameDbObject(newIdentity));
    }
    // reserve address for this identity
    await this.put(mapAddressDbObject(newIdentity));

    // create identity
    await this.put(dbItem);

    return newIdentity;
  }

  public async createProof(input: CreateProofInput): Promise<ProofRecord> {
    const createdAt = new Date(Date.now()).toISOString();

    const newProof = {
      proofType: input.proofType,
      uuid: input.uuid,
      value: input.value,
      createdAt,
    };

    const dbItem = mapProofDbObject(newProof);

    await this.put(dbItem);

    return newProof;
  }

  public async getIdentityByUuid(uuid: string): Promise<IdentityRecord> {
    const rawIdentity = await this.get(getIdentityPrimaryKey(uuid)).then(
      result => result.Item as RawIdentityRecord
    );

    if (!rawIdentity) {
      throw new NotFoundError(`Identity was not found.`);
    }

    return parseDbObjectToIdentity(rawIdentity);
  }

  public async getIdentityByAddress(address: string): Promise<IdentityRecord> {
    const rawAddress = await this.get(getAddressPrimaryKey(address)).then(
      result => result.Item as RawAddressRecord
    );

    if (!rawAddress) {
      throw new NotFoundError(`Identity with address ${address} not found.`);
    }

    const record = parseDbObjectToAddress(rawAddress);
    return this.getIdentityByUuid(record.uuid);
  }

  public getIdentityByPublicKey(publicKey: string): Promise<IdentityRecord> {
    const address = deriveAddressFromPubKey(publicKey);
    return this.getIdentityByAddress(address);
  }

  public async getIdentityByUsername(
    username: string
  ): Promise<IdentityRecord> {
    const rawUsername = await this.get(getUsernamePrimaryKey(username)).then(
      result => result.Item as RawAddressRecord
    );

    if (!rawUsername) {
      throw new NotFoundError(`Identity with username ${username} not found.`);
    }

    const record = parseDbObjectToUsername(rawUsername);
    return this.getIdentityByUuid(record.uuid);
  }

  public async updateIdentity(uuid: string, payload: Record<string, any>) {
    // check that identity exists
    await this.getIdentityByUuid(uuid);

    const allowedKeys = ['displayName', 'avatarUrl'];

    // update uuid with new username
    const Key = getIdentityPrimaryKey(uuid);

    const ExpressionAttributeValues = {};
    const ExpressionAttributeNames = {};

    const updates = [];

    Object.keys(payload).forEach(key => {
      if (allowedKeys.indexOf(key) >= 0) {
        ExpressionAttributeValues[`:${key}`] = payload[key];
        ExpressionAttributeNames[`#${key}`] = key;
        updates.push(`#${key} = :${key}`);
      }
    });

    if (!updates.length) {
      throw new Error('Invalid payload.');
    }

    const UpdateExpression = `SET ${updates.join(',')}`;

    const params = {
      TableName: this.table,
      Select: 'ALL_PROJECTED_ATTRIBUTES',
      Key,
      ExpressionAttributeValues,
      ExpressionAttributeNames,
      UpdateExpression,
    };

    await this.update(params);
  }

  public async changeUsername(uuid: string, username: string) {
    const usernameExists = await this.getIdentityByUsername(username).catch(
      () => false
    );

    if (usernameExists) {
      throw new ValidationError('Username is already taken.');
    }

    const identity = await this.getIdentityByUuid(uuid);

    // assign new username to uuid
    await this.put(
      mapUsernameDbObject({
        uuid,
        username,
        createdAt: new Date().toISOString(),
      })
    );

    // update uuid with new username
    const Key = getIdentityPrimaryKey(uuid);

    const ExpressionAttributeValues = {
      ':username': username,
    };
    const ExpressionAttributeNames = {
      '#username': 'username',
    };
    const UpdateExpression = 'SET #username = :username';

    const params = {
      TableName: this.table,
      Select: 'ALL_PROJECTED_ATTRIBUTES',
      Key,
      ExpressionAttributeValues,
      ExpressionAttributeNames,
      UpdateExpression,
    };

    await this.update(params);

    // release old username
    await this.delete(getUsernamePrimaryKey(identity.username));
  }
}

export default IdentityModel;
