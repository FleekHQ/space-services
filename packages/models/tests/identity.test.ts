/* eslint-disable @typescript-eslint/ban-ts-ignore */
import { IdentityModel } from '../src/identity';
import { mapIdentityDbObject } from '../src/identity/access-patterns';
import { Identity } from '../src/identity/types';
import { ValidationError, NotFoundError } from '../src/errors';

const existingIdentityMock: Identity = {
  address: '0xd606f05a2a980f58737aa913553c8d6eac8b',
  publicKey: '67730a6678566ead5911d71304854daddb1fe98a396551a4be01de65da01f3a9',
  username: 'test-user',
  createdAt: '2020-06-18T21:04:48.141Z',
};
describe('IdentityModel', () => {
  let identityModel: IdentityModel;

  beforeEach(() => {
    identityModel = new IdentityModel('test');
  });

  describe('createIdentity', () => {
    it('fails if the username is taken', async () => {
      identityModel.getIdentityByUsername = () =>
        Promise.resolve(existingIdentityMock);
      identityModel.getIdentityByAddress = () =>
        Promise.reject(new NotFoundError(''));

      const testCreateIdentity = () =>
        identityModel.createIdentity(existingIdentityMock);
      await expect(testCreateIdentity()).rejects.toThrow(
        new ValidationError(
          'An identity with the given username already exists'
        )
      );
    });

    it('fails if the address is taken', async () => {
      identityModel.getIdentityByAddress = () =>
        Promise.resolve(existingIdentityMock);
      identityModel.getIdentityByUsername = () =>
        Promise.reject(new NotFoundError(''));

      const testCreateIdentity = () =>
        identityModel.createIdentity(existingIdentityMock);
      await expect(testCreateIdentity()).rejects.toThrow(
        new ValidationError('An identity with the given address already exists')
      );
    });

    it('creates an identity', async () => {
      identityModel.getIdentityByAddress = () =>
        Promise.reject(new NotFoundError(''));
      identityModel.getIdentityByUsername = () =>
        Promise.reject(new NotFoundError(''));
      // @ts-ignore
      identityModel.put = jest.fn(() => Promise.resolve(null));

      await identityModel.createIdentity(existingIdentityMock);

      // @ts-ignore
      expect(identityModel.put).toHaveBeenCalled();
      // @ts-ignore
      const createdIdentity = identityModel.put.mock.calls[0][0];
      delete createdIdentity.createdAt;
      expect(createdIdentity).toMatchSnapshot();
    });
  });

  describe('getIdentityByAddress', () => {
    it('returns an error if the identity cannot be found', async () => {
      // @ts-ignore
      identityModel.query = () => Promise.resolve({ Items: [] });

      const testGetIdentity = () => identityModel.getIdentityByAddress('0x123');
      await expect(testGetIdentity()).rejects.toThrow(
        new NotFoundError('Identity with address 0x123 not found.')
      );
    });

    it('returns an identity if found', async () => {
      // @ts-ignore
      identityModel.query = () =>
        Promise.resolve({ Items: [mapIdentityDbObject(existingIdentityMock)] });

      const testGetIdentity = () => identityModel.getIdentityByAddress('0x123');
      await expect(testGetIdentity()).resolves.toEqual(existingIdentityMock);
    });
  });

  describe('getIdentityByAddress', () => {
    it('returns an error if the identity cannot be found', async () => {
      // @ts-ignore
      identityModel.query = () => Promise.resolve({ Items: [] });

      const testGetIdentity = () => identityModel.getIdentityByAddress('0x123');
      await expect(testGetIdentity()).rejects.toThrow(
        new NotFoundError('Identity with address 0x123 not found.')
      );
    });

    it('returns an identity if found', async () => {
      // @ts-ignore
      identityModel.query = () =>
        Promise.resolve({ Items: [mapIdentityDbObject(existingIdentityMock)] });

      const testGetIdentity = () => identityModel.getIdentityByAddress('0x123');
      await expect(testGetIdentity()).resolves.toEqual(existingIdentityMock);
    });
  });

  describe('getIdentityByUsername', () => {
    it('returns an error if the identity cannot be found', async () => {
      // @ts-ignore
      identityModel.getItem = () => Promise.resolve({});

      const testGetIdentity = () =>
        identityModel.getIdentityByUsername('test-user');
      await expect(testGetIdentity()).rejects.toThrow(
        new NotFoundError('Identity with username test-user not found.')
      );
    });

    it('returns an identity if found', async () => {
      // @ts-ignore
      identityModel.getItem = () =>
        Promise.resolve({ Item: mapIdentityDbObject(existingIdentityMock) });

      const testGetIdentity = () =>
        identityModel.getIdentityByUsername('test-user');
      await expect(testGetIdentity()).resolves.toEqual(existingIdentityMock);
    });
  });
});
