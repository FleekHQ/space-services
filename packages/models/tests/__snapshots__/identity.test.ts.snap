// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentityModel createIdentity creates an identity 1`] = `
Object {
  "address": "0x0b015d4e6b3f0aeea975efa7287d3daadd0c",
  "gs1pk": "identity",
  "gs1sk": "0x0b015d4e6b3f0aeea975efa7287d3daadd0c",
  "pk": "identity",
  "publicKey": "080112200c0f5e55f4f32af31bf39e1a3703adee3d0141f84f752c1a6522fb358f9c3ac8",
  "sk": "test-user",
  "username": "test-user",
}
`;
